### PR TITLE
Change definition of `curvatureMeasure` and add some lemmas.

### PR DESCRIPTION
**Pull request by LorenzoLuccioli**
- Add `ConvexOn.const_mul`.
- Add lemmas about the right and left derivatives of add, constant and linear functions.
- Add lemmas about `rightDerivStieltjes` of add, constant and linear functions.
- Add `StieltjesFunction.measure_zero`.
- Change the definition of `curvatureMeasure`, now the definition does not require the hypothesis that the function is convex, but uses a junk value whenever this is  not true.
- Add lemmas about `curvatureMeasure` of add, constant and linear functions.

### CODE REVIEW

File: TestingLowerBounds/Convex.lean [hunk 1]
```diff
@@ -77,4 +77,10 @@ lemma comp_neg_iff {𝕜 F β : Type*} [LinearOrderedField 𝕜] [AddCommGroup F
   rw [← neg_neg s, ← Function.comp_id f, ← neg_comp_neg, ← Function.comp.assoc]
   exact h.comp_neg
 
+--this can be stated in much greater generality
+lemma const_mul (c : ℝ) : ConvexOn ℝ Set.univ (fun (x : ℝ) ↦ c * x) := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/Convex.lean]:
```suggestion
lemma const_mul_id (c : ℝ) : ConvexOn ℝ Set.univ (fun (x : ℝ) ↦ c * x) := by
```
I would expect `const_mul` to be the statement that `x ↦ c * f x` is convex (for non-negative c). `const_mul_id ` seems more fitting here.

File: TestingLowerBounds/CurvatureMeasure.lean [hunk 2]
```diff
@@ -7,34 +7,89 @@ import TestingLowerBounds.ForMathlib.ByParts
 import TestingLowerBounds.ForMathlib.LeftRightDeriv
 import Mathlib.MeasureTheory.Integral.FundThmCalculus
 import Mathlib.Probability.Notation
+import Mathlib.Analysis.SpecialFunctions.Gamma.BohrMollerup
```
Comment (RemyDegenne) [inline TestingLowerBounds/CurvatureMeasure.lean:11]: That's a very weird import. What requires that?

File: TestingLowerBounds/CurvatureMeasure.lean [hunk 3]
```diff
@@ -7,34 +7,89 @@ import TestingLowerBounds.ForMathlib.ByParts
...
 -- This way we could avoid having to state the convexity every time.
 -- This may be put in some other place, maybe directly in the stieltjes file.
+
+open Classical in
 /-- The curvature measure induced by a convex function. It is defined as the only measure that has
 the right derivative of the function as a CDF. -/
 noncomputable
-def curvatureMeasure {f : ℝ → ℝ} (hf : ConvexOn ℝ univ f) : Measure ℝ :=
-  hf.rightDerivStieltjes.measure
+irreducible_def curvatureMeasure (f : ℝ → ℝ) : Measure ℝ :=
```
Comment (RemyDegenne) [inline TestingLowerBounds/CurvatureMeasure.lean:30]: Can you remove the comments above that ask whether we should redefine this?

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean [hunk 4]
```diff
@@ -60,6 +60,65 @@ lemma leftDeriv_eq_rightDeriv (f : ℝ → ℝ) :
...
+  ext x
+  rw [rightDeriv_def, Pi.zero_apply]
+  exact derivWithin_const x _ c (uniqueDiffWithinAt_Ioi x)
+
+@[simp]
+lemma leftDeriv_const (c : ℝ) : leftDeriv (fun _ ↦ c) = 0 := by
+  simp_rw [leftDeriv_eq_rightDeriv, rightDeriv_const, Pi.zero_apply, neg_zero]
+  rfl
+
+lemma rightDeriv_linear (a : ℝ) : rightDeriv (fun x ↦ a * x) = fun _ ↦ a := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/ForMathlib/LeftRightDeriv.lean]: That should not be a lemma, I think?
You should add a simp lemma that states that `rightDeriv (fun x ↦ a * f x) = a * rightDeriv f`, and another one that proves `rightDeriv (fun x ↦ x) = fun _ ↦ 1`. Then simp should be able to prove this and `rightDeriv_linear` is not needed.

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean [hunk 5]
```diff
@@ -60,6 +60,65 @@ lemma leftDeriv_eq_rightDeriv (f : ℝ → ℝ) :
...
+  change a * derivWithin id (Ioi x) x = a
+  rw [derivWithin_id _ _ (uniqueDiffWithinAt_Ioi x), mul_one]
+
+lemma leftDeriv_linear (a : ℝ) : leftDeriv (fun x ↦ a * x) = fun _ ↦ a := by
+  simp_rw [leftDeriv_eq_rightDeriv, ← neg_mul_comm, rightDeriv_linear]
+  ext; simp
+
+lemma rightDeriv_add {f g : ℝ → ℝ} (hf : ∀ x, DifferentiableWithinAt ℝ f (Ioi x) x)
+    (hg : ∀ x, DifferentiableWithinAt ℝ g (Ioi x) x) :
+    rightDeriv (f + g) = fun x ↦ rightDeriv f x + rightDeriv g x := by
```
Comment (RemyDegenne) [inline TestingLowerBounds/ForMathlib/LeftRightDeriv.lean:134]: You should have a second version of this, in the form
`rightDeriv (fun x ↦ f x + g x) = fun x ↦ rightDeriv f x + rightDeriv g x`.
Then you wont need the `change` lines in all lemmas below.

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean [hunk 6]
```diff
@@ -187,4 +246,33 @@ def rightDerivStieltjes {f : ℝ → ℝ} (hf : ConvexOn ℝ univ f) :
...
+
+lemma rightDerivStieltjes_add {f g : ℝ → ℝ} (hf : ConvexOn ℝ univ f) (hg : ConvexOn ℝ univ g) :
+    rightDerivStieltjes (hf.add hg) = rightDerivStieltjes hf + rightDerivStieltjes hg := by
+  ext x
+  simp_rw [StieltjesFunction.add_apply, rightDerivStieltjes_eq_rightDeriv, rightDeriv_add
+    (fun x ↦ hf.differentiableWithinAt_Ioi x) (fun x ↦ hg.differentiableWithinAt_Ioi x)]
+
+example (hf : ConvexOn ℝ univ f) (c : ℝ) : ConvexOn ℝ univ (fun x ↦ f x + c) := by
+  exact hf.add (convexOn_const c convex_univ)
+
```
Comment (RemyDegenne) [inline TestingLowerBounds/ForMathlib/LeftRightDeriv.lean]:
```suggestion
```

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean [hunk 7]
```diff
@@ -187,4 +246,33 @@ def rightDerivStieltjes {f : ℝ → ℝ} (hf : ConvexOn ℝ univ f) :
...
+lemma rightDerivStieltjes_add_const (hf : ConvexOn ℝ univ f) (c : ℝ) :
+    rightDerivStieltjes (hf.add (convexOn_const c convex_univ)) = rightDerivStieltjes hf := by
+  rw [rightDerivStieltjes_add hf (convexOn_const c convex_univ), rightDerivStieltjes_const,
+    add_zero]
+
+
```
Comment (RemyDegenne) [inline TestingLowerBounds/ForMathlib/LeftRightDeriv.lean]:
```suggestion
```

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean (hunk 4 above)
Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/LeftRightDeriv.lean]:
```lean
@[simp]
lemma rightDeriv_const_mul (a : ℝ) {f : ℝ → ℝ} :
    rightDeriv (fun x ↦ a * f x) = fun x ↦ a * rightDeriv f x := by
  ext x
  by_cases ha : a = 0
  · simp [ha]
  by_cases hfx : DifferentiableWithinAt ℝ f (Ioi x) x
  · simp_rw [rightDeriv_def]
    rw [derivWithin_const_mul (uniqueDiffWithinAt_Ioi x)]
    exact hfx
  · rw [rightDeriv_of_not_differentiableWithinAt hfx, mul_zero,
      rightDeriv_of_not_differentiableWithinAt]
    refine fun h_diff ↦ hfx ?_
    have : f = fun x ↦ a⁻¹ * (a * f x) := by ext; simp [ha]
    rw [this]
    exact h_diff.const_mul _

@[simp]
lemma rightDeriv_id : rightDeriv id = fun _ ↦ 1 := by
  ext x
  rw [rightDeriv_def, derivWithin_id]
  exact uniqueDiffWithinAt_Ioi x

@[simp]
lemma rightDeriv_id' : rightDeriv (fun x ↦ x) = fun _ ↦ 1 := rightDeriv_id
```

Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/LeftRightDeriv.lean]: those might be useful as well:
```lean
@[simp]
lemma rightDeriv_neg {f : ℝ → ℝ} : rightDeriv (fun x ↦ - f x) = fun x ↦ - rightDeriv f x := by
  have : (fun x ↦ - f x) = fun x ↦ -1 * f x := by simp
  rw [this, rightDeriv_const_mul]
  simp

@[simp]
lemma leftDeriv_neg {f : ℝ → ℝ} : leftDeriv (fun x ↦ - f x) = fun x ↦ - leftDeriv f x := by
  simp [leftDeriv_eq_rightDeriv]

```

File: TestingLowerBounds/ForMathlib/LeftRightDeriv.lean (hunk 5 above)
Comment (RemyDegenne) [reply TestingLowerBounds/ForMathlib/LeftRightDeriv.lean:134]: Also you should write those lemmas for one particular `x` and replace `hf` by `(hf : DifferentiableWithinAt ℝ f (Ioi x) x)`. That way if the differentiability hypothesis is true only on a subset, you can still use the lemma for that subset (that issue came up during the derivAtTop refactor).

File: TestingLowerBounds/CurvatureMeasure.lean (hunk 2 above)
Comment (LorenzoLuccioli) [reply TestingLowerBounds/CurvatureMeasure.lean:11]: I needed that import to have `ConvexOn.add_const`, but I noticed that just 5 days ago it was moved to `Mathlib.Analysis.Convex.Function`, which makes much more sense and is already imported. I will add a comment to remove this import at the next mathlib bump.

**Review by RemyDegenne — APPROVED** [review]

**Pull request merged**